### DISCUSSION TITLE
Stereographic Projection

### DISCUSSION THREAD

**Discussion by offsky** [cosinekitty/astronomy#130] [Show and tell]
One thing I wanted to do with my software project was to draw a sky chart showing the positions of the planets.  From my research, the typical way to do this is to translate the celestial sphere onto a plane using a [Stereographic Projection](https://en.wikipedia.org/wiki/Stereographic_projection).  I figured out the math, so I thought I would share it here in case it's useful to someone else. This bit of Javascript code turns altitude/azimuth into x/y cartesian coordinates that you can then easily draw. The radius variable allows you to scale the x/y coordinates to fit within whatever circle you want.

```
var alt = (alt/57.29577951); //Decimal to radians
var azi = (azi/57.29577951); //Decimal to radians
var D = (2 * radius * radius * Math.cos(alt))/(radius * Math.sin(alt) + radius);
var x = Math.sin(azi) * D;
var y = Math.cos(azi) * D;
```

**Comment by cosinekitty** [comment]
Cool! Thanks for sharing!

**Reply by mayankkalbhor** [reply]
@offsky Do you have a codepen showing that

**Reply by offsky** [reply]
No, but you can see it in action on my website. Its the last dial on this page: https://www.complication.watch

**Reply by mayankkalbhor** [reply]
@offsky This looks great. Cant we use it as an iframe on our website

**Reply by offsky** [reply]
Please don't iframe my website.  You can add a link to my website if you want. 

If you want to add something like this to your website, I encourage you to read the excellent documentation for Astronomy Engine and build it yourself using my 5 line hint above. Its not that difficult to do :)

**Reply by mayankkalbhor** [reply]
Sure thanks @offsky I will work on it.

**Reply by mayankkalbhor** [reply]
@offsky . Seems the website is not working properly. I checked in the console it showing this error and hence the clock faces are not loading

Uncaught ReferenceError: yearLen is not defined

**Reply by offsky** [reply]
I was not expecting to get a bug report on someone elses project, but I appreciate it 😄 It was a leap year bug (2024) that is now fixed. Thanks!

**Comment by matheo** [comment]
Awesome skills @offsky
I imagine there's no way to access the source code?

**Reply by offsky** [reply]
I dont have anything sharable right now, but the meat and potatoes is already posted above. All you need to do is ask Astronomy Engine for the coordinates of the planets. Plug that into the code I shared, and then you have your X/Y where you can draw a circle for the planet.    If you want to draw the constellations too, find a list of the stars and their coordinates and repeat for each star. Boom.

Here are some references where you can learn all about it: 
https://astronomy.stackexchange.com/questions/18219/map-projections-used-for-star-maps
https://en.wikipedia.org/wiki/Stereographic_projection
https://pro.arcgis.com/en/pro-app/latest/help/mapping/properties/orthographic.htm
https://astronomy.stackexchange.com/questions/35882/how-to-make-projection-from-altitude-and-azimuth-to-screen-with-screen-coordinat?utm_source=pocket_mylist
https://astronomy.stackexchange.com/questions/35864/how-to-make-a-star-chart-from-entries-in-a-bright-star-catalogue?utm_source=pocket_mylist
http://www2.arnes.si/~gljsentvid10/starmap.html
